SETUP/upgrade/13/20190929_update_user_project_info.php

```diff
@@ -0,0 +1,111 @@
+<?php
+
+$relPath='../../../pinc/';
+include_once($relPath.'base.inc');
+include($relPath.'udb_user.php');
+
+header('Content-type: text/plain');
+
+// ------------------------------------------------------------
+
+echo "Updating user_project_info to set t_latest_page_event=0 for projects \n";
+echo "where the user has only checked out or returnToRound a page...\n\n";
+
+// Load a list of all users
+$sql = "
+    SELECT username
+    FROM users
+    ORDER BY username
+";
+
+$result = mysqli_query(DPDatabase::get_connection(), $sql)
+    or die( mysqli_error(DPDatabase::get_connection()) );
+
+$users = [];
+while($row = mysqli_fetch_assoc($result))
+{
+    $users[] = $row['username'];
+}
+mysqli_free_result($result);
+
+// Loop though all users getting their user_project_info records to see what
+// projects they have "worked on" (ie: t_latest_page_event > 0)
+foreach($users as $username)
+{
+    echo "Processing $username...\n";
+
+    // load all projects this user has "worked on"
+    $sql = "
+        SELECT projectid
+        FROM user_project_info
+        WHERE
+            username='$username'
+            AND t_latest_page_event > 0
+    ";
+
+    $result = mysqli_query(DPDatabase::get_connection(), $sql)
+        or die( mysqli_error(DPDatabase::get_connection()) );
+
+    $projectids = [];
+    while($row = mysqli_fetch_assoc($result))
+    {
+        $projectids[] = $row['projectid'];
+    }
+    mysqli_free_result($result);
+
+    // for each project, get unique page_events for the project, if the only
+    // two kinds of event_types are 'checkedout' and 'returnToRound' set
+    // user_project_info.t_latest_page_event=0 for that (username, projectid).
+    foreach($projectids as $projectid)
+    {
+        // Process both the main page_events table and the archived
+        // page_events table.
+        $current_events = load_events_from_table("page_events", $username, $projectid);
+        $archived_events = load_events_from_table("$archive_db_name.page_events", $username, $projectid);
+        $event_types = array_unique(array_merge($current_events, $archived_events));
+
+        // see if the only event_types are checkout and returnToRound
+        if(array_diff($event_types, ['checkout', 'returnToRound']) === [])
+        {
+            echo "    setting t_latest_page_event=0 for project $projectid\n";
+            $sql = "
+                UPDATE user_project_info
+                SET t_latest_page_event=0
+                WHERE
+                    username='$username'
+                    AND projectid='$projectid';
+            ";
+            $result = mysqli_query(DPDatabase::get_connection(), $sql)
+                or die( mysqli_error(DPDatabase::get_connection()) );
+        }
+    }
+}
+
+// ------------------------------------------------------------
+
+echo "\nDone!\n";
+
+// ------------------------------------------------------------
+
+function load_events_from_table($table, $username, $projectid)
```

Comment:
```suggestion
function get_distinct_event_types($table, $username, $projectid)
```
("from_table" puts undue emphasis on the table, when "for_user" and "for_project" are just as important. Might as well leave them all implicit in the function name.)